Add carbon emissions to statistics module

gincrement opened the pull request:
Closes #520 

## Changes proposed in this Pull Request


## Checklist

- [X] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [X] Unit tests for new features were added (if applicable).
- [ ] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [X] I consent to the release of this PR's code under the MIT license.

gincrement committed:
add carbon emissions to statistics module (issue #520)

gincrement committed:
adjust statistic test for emissions

gincrement committed:
remove print from test emissions function

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lkstrp reviewed (changes requested):
Thanks for another PR @gincrement !
Could you have a look on all failing checks?

FabianHofmann commented on pypsa/statistics/expressions.py:
```diff
@@ -48,6 +48,12 @@ def port_efficiency(
+    if dynamic:
+        try:
+            etas = pd.Series(n.static(c_name).efficiency, index=n.static(c_name).index)
+        except:
```
I have the feeling that this is already covered below? 

if dynamic is True, the code below will try using a time-dependent efficiency (according to the port). else the static efficiency column is used with a fallback of 1.

gincrement replied:
I tested it, and the code is necessary.

gincrement replied:
removed the try/except construct; but the code is necessary

FabianHofmann reviewed (changes requested):
nice idea to add this!

fneum reviewed (changes requested):
Looks mostly good apart from the comments raised already and one further question from me:

Is it really necessary to use the absolute value function for `get_operation()` and `efficiency`? Imagine the case where you're modelling an integrated BECCS power plant (without modelling the CO2 carrier specifically). Would the function then show net-negative emissions? Perhaps add a test for that case.

gincrement commented on pypsa/statistics/expressions.py:
```diff
@@ -1049,3 +1056,64 @@ def market_value(
+            # n.generators_t.p / n.generators.efficiency * n.generators.carrier.map(n.carriers.co2_emissions)
+            # p / n.generators.efficiency * n.generators.carrier.map(n.carriers.co2_emissions)
+            efficiency = port_efficiency(n, c, port=port, dynamic=True)
+            if not at_port:
```
adjusted accordinlgy

gincrement commented on test/test_statistics.py:
```diff
@@ -261,3 +261,14 @@ def test_parameters(ac_dc_network_r):
+    n.optimize(n.snapshots)
+    # n.statistics().round(1).T
+    # assert reloaded.meta == scipy_network.meta
+    assert (
```
added this additional test

gincrement committed:
Add carbon emissions to statistics module #1149 adjustments

gincrement committed:
Merge branch 'carbon_emission_stats' of https://github.com/gincrement/pypsa into carbon_emission_stats

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lkstrp commented on pypsa/statistics/expressions.py:
```diff
@@ -48,6 +48,12 @@ def port_efficiency(
     ones = pd.Series(1, index=n.static(c_name).index)
+    if dynamic:
+        if "efficiency" in n.static(c_name).columns:
+            etas = pd.Series(n.static(c_name).efficiency, index=n.static(c_name).index)
```
```suggestion
            etas = n.static(c_name).efficiency
```

gincrement committed:
Add carbon emissions to statistics module #932 adj.

gincrement committed:
Merge branch 'carbon_emission_stats' of https://github.com/gincrement/pypsa into carbon_emission_stats

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

FabianHofmann commented on pypsa/statistics/expressions.py:
```diff
@@ -47,7 +47,14 @@ def get_weightings(n: Network, c: str) -> pd.Series:
     n: Network, c_name: str, port: str = "", dynamic: bool = False
 ) -> pd.Series | pd.DataFrame:
-    ones = pd.Series(1, index=n.static(c_name).index)
+    if dynamic:
+        if "efficiency" in n.static(c_name).columns:
+            etas = n.static(c_name).efficiency
+        else:
+            etas = pd.Series(1.0, index=n.static(c_name).index)
+        ones = etas
+    else:
+        ones = pd.Series(1, index=n.static(c_name).index)
```
Ah I think I get the rationale behind it now, you need the efficiencies for generators in order take into account the primary energy use. We just need to be very careful here as for one-port component this function always used to refer to the output energy at a port (not the primary energy input). For understanding this better, it the new emission mainly tailored to generators? or also links might have an carbon output modeled as a secondary output?

gincrement replied:
My understanding is that global constraints are designed for generators and generators are the mean to bring fuels/emissions into the system. Correct?

gincrement committed:
Merge branch 'master' into carbon_emission_stats

FabianHofmann committed:
Merge branch 'master' into pr/gincrement/1149-1

FabianHofmann committed:
remove unwanted addition from merge

FabianHofmann committed:
revert component default

FabianHofmann commented:
after notifying @gincrement that changing the `port_efficiency` function would lead to quite some problems in the other statistic functions, I added the efficiency calculation to the carbon_emission function directly. Note that the current implementation does only support generators, it does not include Stores and StorageUnits as imposed by the corresponding constraint. I honestly don't know if we can/should include them. probably it would be nice and add them but I had to think about it more...

FabianHofmann committed:
Merge branch 'master' into carbon_emission_stats

fneum reviewed (changes requested):
I think it should match the corresponding global constraint:

https://pypsa.readthedocs.io/en/latest/user-guide/optimal-power-flow.html#primary-energy

e.g. looking at the net supply of the store/storage unit between initial and final state of charge.

Just returning the carbon emissions of the generators could be confusuing and would not be aligned with the global constraint. That should be avoided.

Even though the statistics function prints a warning, I think the stores and storage units implementation needs to be included before merging.

gincrement committed:
Update expressions.py

removed abs from efficiency and get_operation and increased coverate to all technologies

gincrement committed:
Update test_statistics.py

added co2 carrier, bus and load to test negative and positive flows

fneum committed:
Delete qq

fneum reviewed (changes requested):
Another unit test would be good where emissions come from the net difference in state of charge of a Store / StorageUnit. Then, I'm fine for this to be merged. 

See

https://github.com/PyPSA/PyPSA/blob/5031132a952d4c708ca1d304d219b906fa0524e7/pypsa/optimization/global_constraints.py#L272-L480

gincrement committed:
adjust test case
